pkg/objstore/swift/swift.go

```diff
@@ -9,69 +9,177 @@ import (
 	"github.com/pkg/errors"
 	"gopkg.in/yaml.v2"
 
 	"github.com/thanos-io/thanos/pkg/objstore"
 )
 
 // DirDelim is the delimiter used to model a directory structure in an object store bucket.
-const DirDelim = "/"
-
-type SwiftConfig struct {
-	AuthUrl           string `yaml:"auth_url"`
-	Username          string `yaml:"username"`
-	UserDomainName    string `yaml:"user_domain_name"`
-	UserDomainID      string `yaml:"user_domain_id"`
-	UserId            string `yaml:"user_id"`
-	Password          string `yaml:"password"`
-	DomainId          string `yaml:"domain_id"`
-	DomainName        string `yaml:"domain_name"`
-	ProjectID         string `yaml:"project_id"`
-	ProjectName       string `yaml:"project_name"`
-	ProjectDomainID   string `yaml:"project_domain_id"`
-	ProjectDomainName string `yaml:"project_domain_name"`
-	RegionName        string `yaml:"region_name"`
-	ContainerName     string `yaml:"container_name"`
+const (
+	DirDelim    = '/'
+	SegmentsDir = "segments/"
+)
+
+var DefaultConfig = Config{
+	AuthVersion:    0, // Means autodetect of the auth API version by the library.
+	ChunkSize:      1024 * 1024 * 1024,
+	Retries:        3,
+	ConnectTimeout: model.Duration(10 * time.Second),
+	Timeout:        model.Duration(5 * time.Minute),
+}
+
+type Config struct {
+	AuthVersion          int            `yaml:"auth_version"`
+	AuthUrl              string         `yaml:"auth_url"`
+	Username             string         `yaml:"username"`
+	UserDomainName       string         `yaml:"user_domain_name"`
+	UserDomainID         string         `yaml:"user_domain_id"`
+	UserId               string         `yaml:"user_id"`
+	Password             string         `yaml:"password"`
+	DomainId             string         `yaml:"domain_id"`
+	DomainName           string         `yaml:"domain_name"`
+	ProjectID            string         `yaml:"project_id"`
+	ProjectName          string         `yaml:"project_name"`
+	ProjectDomainID      string         `yaml:"project_domain_id"`
+	ProjectDomainName    string         `yaml:"project_domain_name"`
+	RegionName           string         `yaml:"region_name"`
+	ContainerName        string         `yaml:"container_name"`
+	ChunkSize            int64          `yaml:"large_object_chunk_size"`
+	SegmentContainerName string         `yaml:"large_object_segments_container_name"`
+	Retries              int            `yaml:"retries"`
+	ConnectTimeout       model.Duration `yaml:"connect_timeout"`
+	Timeout              model.Duration `yaml:"timeout"`
+}
+
+func parseConfig(conf []byte) (*Config, error) {
+	sc := DefaultConfig
+	err := yaml.UnmarshalStrict(conf, &sc)
+	return &sc, err
+}
+
+func configFromEnv() (*Config, error) {
+	c := swift.Connection{}
+	if err := c.ApplyEnvironment(); err != nil {
+		return nil, err
+	}
+
+	config := Config{
+		AuthVersion:          c.AuthVersion,
+		AuthUrl:              c.AuthUrl,
+		Password:             c.ApiKey,
+		Username:             c.UserName,
+		UserId:               c.UserId,
```

Comment:
Very minor thing but in this `Config` we should set `ChunkSize` to `DefaultConfig.ChunkSize`, right? Otherwise it is always zero if `SWIFT_CHUNK_SIZE` is empty.

Reply:
Right, I missed that, thanks!

Reply:
Added here https://github.com/thanos-io/thanos/pull/2732/commits/7a51f585a7184cc9a0c5a73ffcd77b23896da026